fix: Fix default values and dictionary values for session replay options

philprime opened the pull request:
# :scroll Description

This commit introduces several enhancements to the SentryReplayOptions class, including new properties for better configuration and handling of session replay options. Key changes include:

- Added internal methods for converting raw values to SentryReplayQuality.
- Introduced new properties for bitrate and size scale calculations.
- Updated initializers to support dictionary-based configuration.
- Added comprehensive unit tests for SentryReplayOptions and related classes, ensuring default values and custom configurations are correctly handled.
- Introduced new test files for Objective-C compatibility and SwiftUI integration.

Covered cases:
- Init without args in Swift
- Init with all args in Swift
- Init with some args ommitted in Swift (fallback to default value)
- No nullable args
- Init without args in Objective-C
- Init with all args in Objective-C
- Init with missing value in dictionary, fallback to default value

## :bulb: Motivation and Context

- The default values are inconsistent (i.e. `enableViewRendererV2` was not enabled by default everywhere)
- The default values set in the Swift initializer made an initializer without arguments unavailable to Objective-C, and the replacement fix (empty init) was setting default values inconsistently.

## :green_heart: How did you test it?

## :pencil: Checklist

You have to check all boxes before merging:

- [X] I added tests to verify the changes.
- [X] No new PII added or SDK only sends newly added PII if `sendDefaultPII` is enabled.
- [X] I updated the docs if needed.
- [X] I updated the wizard if needed.
- [X] Review from the native team if needed.
- [X] No breaking change or entry added to the changelog.
- [X] No breaking change for hybrid SDKs or communicated to hybrid SDKs.

philprime commented:
cc @kahest so you know this PR exists.

armcknight commented on Sources/Swift/Integrations/SessionReplay/SentryReplayOptions.swift:
```diff
@@ -216,39 +278,108 @@ public class SentryReplayOptions: NSObject, SentryRedactOptions {
+        // Note for future maintainers:
+        // - This initializer is publicly available for Swift, but not for Objective-C.
+        // - Each parameter has a default value, so the parameter can be omitted.
+        // - The values are not optional, because SDK users should not be able to set them to nil.
```
The fact that the declarations are not optional should be sufficient, no need to restate it here. Additionally, if they ever become optional for some reason, there is no automatic way to know this comment needs to be changed, and so could become a lie in the future. I would just remove it.

philprime replied:
Code is changed easily, this is context to make maintainers aware we decided on purpose not to make it optional, and have a default fallback in the initializer body.

armcknight commented on Sources/Swift/Integrations/SessionReplay/SentryReplayOptions.swift:
```diff
@@ -216,39 +278,108 @@ public class SentryReplayOptions: NSObject, SentryRedactOptions {
+        enableFastViewRendering: Bool = false
+    ) {
+        // Note for future maintainers:
+        // - This initializer is publicly available for Swift, but not for Objective-C.
```
I think it is common knowledge that Swift initializers with default values are not available to ObjC, we don't need to write this in every spot.

Additionally, if it is indeed intended that we allow consumers to instantiate an instance of SentryReplayOptions from objc, we should code that in the iOS-ObjectiveC sample app to avoid breaking things in the future.

philprime replied:
Yes, common knowledge if you have background in using Objective-C. But in my case I didn't know why the method was there in the first place, because Sesssion Replay is mostly written and used in Swift (in the SDK), but then I realized the constructor is there because of `SentryOptions.m`.

philipphofmann replied:
I think it would be good to comment on why it's not available in ObjC.

armcknight commented on Sources/Swift/Integrations/SessionReplay/SentryReplayOptions.swift:
```diff
@@ -153,60 +175,100 @@ public class SentryReplayOptions: NSObject, SentryRedactOptions {
      *               view renderer, please report the issue on [GitHub](https://github.com/getsentry/sentry-cocoa). Eventually, we will
      *               mark this feature as stable and remove the experimental flag, but will keep it disabled by default.
      */
-    public var enableFastViewRendering = false
+    public var enableFastViewRendering: Bool
 
     /**
      * Defines the quality of the session replay.
      * Higher bit rates better quality, but also bigger files to transfer.
      */
-    var replayBitRate: Int {
-        quality.rawValue * 20_000 + 20_000
+    internal var replayBitRate: Int {
```
Why add all these? Isn't `internal` the default?

philprime replied:
My intention was to be explicit what's public facing and what not, so that any future maintainer visiting this property will reconsider if making it public is necessary. If no visibility is defined, some might be careless and think something like: "Oh we missed adding a `public` here, that's why it's not available".

philipphofmann replied:
As internal is the default, it's a bit redundant, but I don't care too much about that.

philprime replied:
Removed the internal

armcknight reviewed (commented):
Is this change entirely motivated by the inconsistency in the default value being set for `enableViewRendererV2`? Personally, I find these changes confusing and potentially dangerous. We cannot rely on comments to enforce consistency in the business logic, those kinds of constraints need to be coded. Yes, you can catch them in unit tests, but this introduces almost 1,000 lines of test; that is also code that is prone to errors, requires its own maintenance, and slows down future changes. A lot of that checking can be done by the compiler if we keep the target code simpler. And we wouldn't need the hacks to avoid recursive convenience init calls.

My approach here would probably involve deduplicating magic values that are used as defaults in variable declarations and default parameter values into static variables or even into a structure called Defaults. One source of truth. Here's what that would look like: https://github.com/getsentry/sentry-cocoa/pull/5214. A lot of your tests might still provide value so I don't necessarily think you need to throw all that effort away.

It's unfortunate that this class was ever written with publicly available initializers and mutable properties, as this is the kind of mess it leads to. Better would have been to use a builder pattern instead of offering multiple types of initializers. We use that elsewhere so I'm not sure why we didn't here. Sorry you have to deal with that now 😭

philprime commented:
Thanks for the feedback @armcknight, I already had a discussion with @philipphofmann yesterday, that this situation is quite complex and due the some of these initializers being public and already shipped to production made it even harder and frustrating to find a solution that covers all. 

> Is this change entirely motivated by the inconsistency in the default value being set for enableViewRendererV2? Personally, I find these changes confusing and potentially dangerous. We cannot rely on comments to enforce consistency in the business logic, those kinds of constraints need to be coded. Yes, you can catch them in unit tests, but this introduces almost 1,000 lines of test; that is also code that is prone to errors, requires its own maintenance, and slows down future changes. A lot of that checking can be done by the compiler if we keep the target code simpler. And we wouldn't need the hacks to avoid recursive convenience init calls.

This is the second time that a default value was only partially updated and therefore not consistent, so it must be covered somehow. I agree with most of your statements, these options are just so badly designed that any change can cause public facing issues. IMHO some of the initializers should not have been public in the first place.

> My approach here would probably involve deduplicating magic values that are used as defaults in variable declarations and default parameter values into static variables or even into a structure called Defaults. One source of truth. Here's what that would look like: https://github.com/getsentry/sentry-cocoa/pull/5214. A lot of your tests might still provide value so I don't necessarily think you need to throw all that effort away.

Yes, I had this one approach in my drafts as well but it had two additional downsides:

1. We would introduce another public type for the constants, something I wanted to avoid because SDK users should not care about the default values like that (we can not use an internal type, the type must be public).
2. A person looking at the Xcode documentation of the initializer will not see anymore what the default value is:

<img width="555" alt="image" src="https://github.com/user-attachments/assets/44a15f7a-6b2a-40d1-9766-d0b8bd8174c8" />

> It's unfortunate that this class was ever written with publicly available initializers and mutable properties, as this is the kind of mess it leads to. Better would have been to use a builder pattern instead of offering multiple types of initializers. We use that elsewhere so I'm not sure why we didn't here. Sorry you have to deal with that now 😭

Yes, this is terrible to maintain, but it is what is now. I added [a note](https://github.com/getsentry/sentry-cocoa/discussions/2946#discussioncomment-13085656) for v9

philprime commented on Sources/Swift/Integrations/SessionReplay/SentryReplayOptions.swift:
```diff
@@ -153,60 +175,100 @@ public class SentryReplayOptions: NSObject, SentryRedactOptions {
+        self.init(
+            // We need to set at least one parameter to avoid the initializer to call itself.
+            //
+            // MAKE SURE THIS DEFAULT VALUE IS ALIGNED WITH THE DEFAULT VALUE IN THE INITIALIZER BELOW.
+            sessionSampleRate: 0
```
I agree with comments not being good quality control. If we want a static variable in the constructor, it must be public, see [my PR comment](https://github.com/getsentry/sentry-cocoa/pull/5210#issuecomment-2865523968).

philipphofmann commented:
I don't believe this code will be terrible to maintain, because we're not going to change that code often, as it's a public API. The tests can only change when the requirements change, which should only be the case in the next major release. As these APIs are a bit of a mess and we can't fix that now, let's agree on a somewhat OK solution and fix it later.

philipphofmann commented on Sources/Swift/Integrations/SessionReplay/SentryReplayOptions.swift:
```diff
@@ -216,39 +278,108 @@ public class SentryReplayOptions: NSObject, SentryRedactOptions {
+    ) {
+        // Note for future maintainers:
+        // - This initializer is publicly available for Swift, but not for Objective-C.
+        // - Each parameter has a default value, so the parameter can be omitted.
```
I agree with both of you. If it helps you @philprime for future context, we can keep it. It doesn't hurt much.

philprime replied:
I want to keep this, added a note related to unavailability in ObjC so the comment is a comparision between Swift and ObjC and not just an explanaiton of the Swift compiler.

philipphofmann commented on Sources/Swift/Integrations/SessionReplay/SentryReplayOptions.swift:
```diff
@@ -216,39 +278,108 @@ public class SentryReplayOptions: NSObject, SentryRedactOptions {
+        // - Each parameter has a default value, so the parameter can be omitted.
+        // - The values are not optional, because SDK users should not be able to set them to nil.
+
+        // WHEN CHANGING ANY DEFAULT VALUE IN THE INITIALIZER, UPDATE THE PROPERTY DOCUMENTATION
```
`l`: Maybe a bit more specific.
```suggestion
        // WHEN CHANGING ANY DEFAULT VALUE IN THE INITIALIZER, UPDATE THE PROPERTY CODE DOCUMENTATION ABOVE
```

philipphofmann commented on Sources/Swift/Integrations/SessionReplay/SentryReplayOptions.swift:
```diff
@@ -216,39 +278,108 @@ public class SentryReplayOptions: NSObject, SentryRedactOptions {
         self.enableViewRendererV2 = enableViewRendererV2
         self.enableFastViewRendering = enableFastViewRendering
+
+        // These are the default values for the properties.
+        // Do not add any default values at the property declaration, because they will be overridden by the initializer.
```
`m`: Now looking at this again, I'm reconsidering my statement yesterday. I think having the default values in one place as suggested in https://github.com/getsentry/sentry-cocoa/pull/5214, which has the downside of not showing the default values in the code docs, is easier to maintain. The current approach can break easily, but we're also not going to change this often. We can work around the code doc default value problem, by adding code docs for the initializers and the properties. The SentryOptions are written in ObjC and you can also only get the default value by looking at the code docs. I have no strong opinion on this one. I let you decide @philprime.

philprime replied:
We are trading maintainability for usability. I'll create a well-documented defaults type so that users can also easily understand the values. I will NOT add the default values to the code documentation for this initializer, because then we are back at the original issue of having multiple fields to maintain.

philprime replied:
I decided to add a comment to each initializer and property pointing to the default values type.

philipphofmann commented on Tests/SentrySwiftUITests/PreviewRedactOptionsTests.swift:
```diff
@@ -0,0 +1,140 @@
+        XCTAssertFalse(options.maskAllText)
+        XCTAssertFalse(options.maskAllImages)
+        XCTAssertEqual(options.maskedViewClasses.count, 1)
+        for maskedViewClass in options.maskedViewClasses {
+            XCTAssertEqual(String(describing: maskedViewClass), String(describing: UIView.self))
+        }
```
`l`: Maybe we can use `assertArrayEquals` or some other helper method to avoid having to write this assertion code multiple times.

philipphofmann commented on Tests/SentryTests/Integrations/SessionReplay/SentryReplayOptionsObjcTests.m:
```diff
@@ -0,0 +1,41 @@
+
+- (void)testInit_withoutArguments_shouldUseDefaults
+{
+    // Note for future maintainers:
```
`l`: Every note is for future maintainers 😅 . No need to point that out.
```suggestion
```

philipphofmann commented on Tests/SentryTests/Integrations/SessionReplay/SentryReplayOptionsTests.swift:
```diff
@@ -3,34 +3,217 @@ import Foundation
+        XCTAssertEqual(options.maximumDuration, 60 * 60)
+    }
+
+    func testInit_sessionSampleRateOmitted_shouldUseDefaultValues() {
```
`l`: Can't we reduce the test cases by omitting all except one parameter in one test, and then omitting just this parameter in another test. We basically just need two test cases. So basically something like this

Test omitting all other than `sessionSampleRate`
```
let options = SentryReplayOptions(
            sessionSampleRate: 0.5,
        )
```

Tests omitting `sessionSampleRate`
```
let options = SentryReplayOptions(
           
            onErrorSampleRate: 0.8,
            maskAllText: false,
            maskAllImages: false,
            enableViewRendererV2: false,
            enableFastViewRendering: true
        )
```

philprime replied:
I believe we should test them omitted individually because removing most of the defaults could easily cause a different initializer being tested instead, e.g. removing all would cause the default initializer to be triggered instead of using all default values --> the original problem.

Relying on at least one value being set could work, which was the previous attempt aswell.

At least I can remove the assertions for unrelated fields, so we only assert the default value of each omitted field

philipphofmann replied:
Let's keep them then.

philipphofmann reviewed (approved):
LGTM, we need so many tests because the logic is so complex. Without all these tests,  we can't be sure everything works. We can reduce the number of tests to omit some of the parameters, but the rest is required, IMO. FYI @armcknight.

Thanks for adding all the tests, @philprime

philipphofmann commented:
> It's unfortunate that this class was ever written with publicly available initializers and mutable properties, as this is the kind of mess it leads to. Better would have been to use a builder pattern instead of offering multiple types of initializers. We use that elsewhere so I'm not sure why we didn't here. Sorry you have to deal with that now 😭

And yes, that's the root cause. We should fix this in V9.

philprime commented on Sources/Swift/Integrations/SessionReplay/SentryReplayOptions.swift:
```diff
@@ -216,39 +278,108 @@ public class SentryReplayOptions: NSObject, SentryRedactOptions {
+        onErrorSampleRate: Float = 0,
+        maskAllText: Bool = true,
+        maskAllImages: Bool = true,
+        enableViewRendererV2: Bool = true,
```
Thank you for testing!

philprime commented on CHANGELOG.md:
```diff
@@ -5,6 +5,7 @@
 ### Fixes
 
 - Detect AppHangsV2 when tracing not enabled (#5184)
+- Fix default and dictionary values for session replay options (#5210)
```
I added two entries, one for the fix of unaligned default values and one for the improvements regarding the dictionary initializer for Hybrid SDKs

philprime commented:
Based on your review feedback I refactored the initializers to use the default values in the new public type `SessionReplayOptions.DefaultValues`. I added a reference to all of the initializers and properties so SDK users can find more details

antonis reviewed (approved):
Thank you for all the changes @philprime 🙇 
The changes on this PR LGTM and worked as expected [with RN](https://github.com/getsentry/sentry-react-native/pull/4815) in my tests 🚀

philipphofmann commented on CHANGELOG.md:
```diff
@@ -5,6 +5,11 @@
 ### Fixes
 
 - Detect AppHangsV2 when tracing not enabled (#5184)
+- Add shared default values for session replay options initializers (#5210)
```
`m`: Can we point out which default values we fixed, please?
```suggestion
- Fix default values for session replay options (#5210) TODO add the fixed ones.
```

philprime replied:
I adapted it, it's only the `enableViewRendererV2` which changed in the initializer.

philipphofmann commented on Sources/SentrySwiftUI/Preview/PreviewRedactOptions.swift:
```diff
@@ -1,14 +1,66 @@
     public let maskAllText: Bool
+
+    /**
+     * Flag to redact all non-bundled image in the app by drawing a black rectangle over it.
```
`l`: A typo, I believel
```suggestion
     * Flag to redact all non-bundled images in the app by drawing a black rectangle over them.
```

philipphofmann commented on Sources/SentrySwiftUI/Preview/PreviewRedactOptions.swift:
```diff
@@ -1,14 +1,66 @@
     public let unmaskedViewClasses: [AnyClass]
+    
+    /**
+     * Enables the up to 5x faster new view renderer used by the Session Replay integration.
```
`l`: It soon won't be new, so we can ditch it.
```suggestion
     * Enables the up to 5x faster view renderer used by the Session Replay integration.
```

philipphofmann commented on Sources/SentrySwiftUI/Preview/PreviewRedactOptions.swift:
```diff
@@ -1,14 +1,66 @@
+     *
+     * - Parameters:
+     *   - maskAllText: Flag to redact all text in the app by drawing a black rectangle over it.
+     *   - maskAllImages: Flag to redact all non-bundled image in the app by drawing a black rectangle over it.
```
`l`: Same typo here.
```suggestion
     *   - maskAllImages: Flag to redact all non-bundled images in the app by drawing a black rectangle over them.
```

philipphofmann commented on Sources/SentrySwiftUI/Preview/PreviewRedactOptions.swift:
```diff
@@ -1,14 +1,66 @@
+     *   - maskAllImages: Flag to redact all non-bundled image in the app by drawing a black rectangle over it.
+     *   - maskedViewClasses: The classes of views to mask.
+     *   - unmaskedViewClasses: The classes of views to exclude from masking.
+     *   - enableViewRendererV2: Enables the up to 5x faster new view renderer used by the Session Replay integration.
```
`l`: Also ditch new.
```suggestion
     *   - enableViewRendererV2: Enables the up to 5x faster view renderer used by the Session Replay integration.
```

philipphofmann commented on Sources/Swift/Integrations/SessionReplay/SentryReplayOptions.swift:
```diff
@@ -152,103 +207,206 @@ public class SentryReplayOptions: NSObject, SentryRedactOptions {
+     * Initializes a new instance of ``SentryReplayOptions`` with the specified parameters.
+     *
+     * - Parameters:
+     *   - sessionSampleRate: Indicates the percentages in which replay for the session will be created.
+     *   - onErrorSampleRate: Indicates the percentages in which a buffer of 30 seconds replay will be send with error events.
```
```suggestion
     *   - sessionSampleRate: Indicates the percentage in which replay for the session will be created.
     *   - onErrorSampleRate: Indicates the percentage in which a buffer of 30 seconds replay will be sent with error events.
```

philipphofmann commented on Sources/Swift/Integrations/SessionReplay/SentryReplayOptions.swift:
```diff
@@ -152,103 +207,206 @@ public class SentryReplayOptions: NSObject, SentryRedactOptions {
+     * Initializes a new instance of ``SentryReplayOptions`` with the specified parameters.
+     *
+     * - Parameters:
+     *   - sessionSampleRate: Indicates the percentages in which replay for the session will be created.
+     *   - onErrorSampleRate: Indicates the percentages in which a buffer of 30 seconds replay will be send with error events.
```
`m`: I'm a bit confused by the wording here. I don't really understand this. Maybe try to rephrase it, please.

philprime replied:
I agree, these were already in the class before I started refactoring. Will replace them.

philipphofmann replied:
> these were already in the class before I started refactoring

That explains it, then 😅 .  I'm used to higher standard comments from you 😄

philprime replied:
I rewrote them, should be more concise now.

philipphofmann commented on Sources/Swift/Integrations/SessionReplay/SentryReplayOptions.swift:
```diff
@@ -152,103 +207,206 @@ public class SentryReplayOptions: NSObject, SentryRedactOptions {
+     *   - sessionSampleRate: Indicates the percentages in which replay for the session will be created.
+     *   - onErrorSampleRate: Indicates the percentages in which a buffer of 30 seconds replay will be send with error events.
+     *   - maskAllText: Flag to redact all text in the app by drawing a black rectangle over it.
+     *   - maskAllImages: Flag to redact all non-bundled image in the app by drawing a black rectangle over it.
```
`l`:
```suggestion
     *   - maskAllImages: Flag to redact all non-bundled images in the app by drawing a black rectangle over them.
```

philipphofmann commented on Sources/Swift/Integrations/SessionReplay/SentryReplayOptions.swift:
```diff
@@ -152,103 +207,206 @@ public class SentryReplayOptions: NSObject, SentryRedactOptions {
+     *   - onErrorSampleRate: Indicates the percentages in which a buffer of 30 seconds replay will be send with error events.
+     *   - maskAllText: Flag to redact all text in the app by drawing a black rectangle over it.
+     *   - maskAllImages: Flag to redact all non-bundled image in the app by drawing a black rectangle over it.
+     *   - enableViewRendererV2: Enables the up to 5x faster new view renderer used by the Session Replay integration.
```
```suggestion
     *   - enableViewRendererV2: Enables the up to 5x faster view renderer used by the Session Replay integration.
```

philipphofmann reviewed (approved):
LGTM, with a couple of suggestions. Thanks

philipphofmann commented on Sources/Swift/Integrations/SessionReplay/SentryReplayOptions.swift:
```diff
@@ -333,12 +339,12 @@ public class SentryReplayOptions: NSObject, SentryRedactOptions {
-     *   - maskAllImages: Flag to redact all non-bundled image in the app by drawing a black rectangle over it.
-     *   - enableViewRendererV2: Enables the up to 5x faster new view renderer used by the Session Replay integration.
-     *   - enableFastViewRendering: Enables faster but incomplete view rendering used by the Session Replay integration.
+     *   - sessionSampleRate: Sample rate used to determine the percentage of replays of sessions that will be uploaded.
+     *   - onErrorSampleRate: Sample rate used to determine the percentage of replays of error events that will be uploaded.
```
Better thanks

philipphofmann reviewed (approved):
Thanks

armcknight commented:
Nice work getting this across the finish line @philprime! 🥇
